Store `DerivationOption` inside `Derivation`, put in JSON format

haenoe opened the pull request:
# Motivation

This opens the door to derivations that *directly* specify their options, rather than "stealing" environment variables to do so.

The A-Term derivation format used on disk didn't change --- we cannot do that for existing derivations, so those derivations will continue to not support separate options. But having a more flexible JSON format opens the door to extending the on-disk format in others, like directly using JSON, using CBOR, etc.

# Context

Fixes #9846

Progress on #9866 (regarding the "magic values" `env` field that are used as options)

Depends on #13263
Depends on #14506

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Ericson2314 committed:
Store `DerivationOption` instead `Derivation`, put in JSON format

This opens the door to derivations that *directly* specify their
options, rather than "stealing" environment variables to do so.

The A-Term derivation format used on disk didn't change --- we cannot do
that for existing derivations, so those derivations will continue to not
support separate options. But having a more flexible JSON format opens
the door to extending the on-disk format in others, like directly using
JSON, using CBOR, etc.

Co-authored-by: HaeNoe <git@haenoe.party>
Co-authored-by: Jonathan Gibbons <jonored@gmail.com>

libstore: introduce `DerivationATerm`

A new data type in `derivation-aterm.{hh,cc}` that corresponds closely
to the on-disk ATerm format: inputs split into sources and derivations
(`FullInputs`), and the environment kept *verbatim* — including the
legacy encodings of the derivation options and the `__json` encoding of
structured attributes. It has no `name` field, since the name is not
part of the format proper (it comes from the store path).

Parsing and printing are now two-staged:

* `DerivationATerm::parse` / `to_string` deal only with the textual
  format.

* `DerivationATerm::elaborate` parses the legacy environment-variable
  encodings into a full `Derivation`; `DerivationATerm::lower` is the
  reverse *lossy projection*, which simply drops the parsed fields.

Rather than re-encoding fields on unparse, `Derivation::unparse` checks
faithfulness by round-tripping: a derivation is representable in the
ATerm format iff `lower(drv).elaborate(store) == drv`, i.e. iff its
parsed fields are in sync with their environment-variable encoding.
Derivations whose options are specified directly (e.g. via newer
formats like JSON) without the legacy environment encoding are thus
not ATerm-representable, and `unparse` throws for them.

The internal `unparseDerivation` no longer knows about structured
attributes; the hash-modulo entry points now merge the `__json`
variable into their masked environments themselves.

And more below!

edolstra commented on src/libstore/derivations.hh:
```diff
@@ -289,6 +362,8 @@ struct BasicDerivation
     StringPairs env;
     std::string name;
 
+    DerivationOptions options;
```
It seems that having this field in `BasicDerivation` "denormalizes" it, since information like the value of `__impureEnvVars` is now stored in both `drv.env` *and* `drv.options.impureEnvVars`, which could get out of sync and would therefore be error-prone to use. Wouldn't it be better to have methods like `BasicDerivation::{get,set}ImpureEnvVars()`?

Ericson2314 replied:
@edolstra the goal is, via the JSON format, to allow that stuff to not exist in the env. Then only the legacy decoding results looks like denormalization — but actually I prefer the perspective that there is no denormalization and rather the legacy format simply lacks the expressive power to set one without setting the other.

Does that make sense? It is some subtle philosophical perspective-shifting.

roberth replied:
Philisophical doesn't do it justice. It's anticipating a design improvement.

Ericson2314 replied:
Yeah philosophical before the new feature (better JSON format *as input*), practical with it.

haenoe commented on src/libstore/derivations.cc:
```diff
@@ -518,6 +518,27 @@ DerivationOptions DerivationOptions::fromEnv(const StringPairs & env)
 
     DerivationOptions defaults = {};
 
+    if (auto structuredAttrs = parsed.getStructuredAttrs()) {
```
Let me know if you think this should rather stay [here](https://github.com/haenoe/nix/blob/6f64154eeac328c87e62c1d8ae3aacd90ab9d5d7/src/libstore/unix/build/local-derivation-goal.cc#L2906) (in `local-derivation-goal.cc`).

haenoe replied:
Reverted this, because the change pollutes the output (as we parse with `DerivationOptions::fromEnv` multiple times while evaluating a derivation.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-06-24-nix-team-meeting-minutes-155/47739/1

Ericson2314 commented on src/libstore/derivation-options.hh:
```diff
@@ -2,11 +2,24 @@
 ///@file
 
 #include <cstdint>
+#include <nlohmann/json.hpp>
```
There is `json_fwd` we should use in headers

Ericson2314 commented on src/libstore/derivation-options.hh:
```diff
@@ -52,6 +65,8 @@ struct DerivationOptions
 
     OutputChecks checksAllOutputs;
 
+    AdditionalAttributes attrs;
+
```
I don't think this should go here, but replacing `env` in `BasicDerivation`

haenoe replied:
Ah yes oopsie, should it still be named `env` though?

Ericson2314 replied:
name doesn't matter, since call sites need to be updated either way.

Yes probably should go there.

haenoe commented:
Long time no see ^^'
Tests will fail, at least on my machine due to the unit test [here](https://github.com/haenoe/nix/blob/b7b3ad5cdd24ec1db161cbf9b4b472229beba34f/src/libstore-tests/derivation-advanced-attrs.cc#L71)
I tried to debug it, but my tired brain is stuck :D Will try again later this week!

Ericson2314 commented on src/libstore-tests/data/derivation/ca/advanced-attributes-defaults.json:
```diff
@@ -15,6 +15,27 @@
   "inputSrcs": [],
   "name": "advanced-attributes-defaults",
+  "options": {
+    "additionalSandboxProfile": "",
+    "allowLocalNetworking": false,
+    "allowSubstitutes": true,
+    "impureEnvVars": [],
+    "impureHostDeps": [],
+    "noChroot": false,
+    "outputChecks": {
+      "forAllOutputs": {
+        "allowedReferences": null,
+        "allowedRequisites": null,
+        "disallowedReferences": [],
+        "disallowedRequisites": [],
+        "ignoreSelfRefs": true
+      }
+    },
+    "passAsFile": [],
+    "preferLocalBuild": false,
+    "requiredSystemFeatures": [],
+    "unsafeDiscardReferences": {}
```
So I would, contra our usual JSON policies, *not* want unused options in here. This is because I want to think of options as an "*extensible* record", where various Nix implementations are e.g. free to do their own custom options under their own prefix, and whenever an unknown option is encountered, as sensible warning is given.

Ericson2314 committed:
WIP only use `FullInputs` for formats

amaanq committed:
libstore: compute the hash modulo via an intermediate derivation

`hashDerivationModulo` used to smuggle the hash-substituted inputs into
`unparse` through an extra parameter, overriding what the derivation
value actually contained. Now the substitution produces a real
intermediate value, `DerivationT<HashModuloInputs>`, and a private
`unparseDerivation` free function renders it.

The public `DerivationT<FullInputs>::unparse` no longer takes a
`maskOutputs` parameter — it always serialises faithfully (no masking).
Output masking is now confined to the private `unparseDerivation`, which
takes `maskOutputs` as a template parameter and `static_assert`s that
`FullInputs` is never masked.

`hashDerivationModuloImpl` is a file-private template on `maskOutputs`.
`processDerivationOutputPaths` instantiates it with `true` (mask outputs
to avoid circularity when computing a derivation's own output paths),
while the public `hashDerivationModulo` and `pathDerivationModulo` use
`false` (outputs visible, for computing a derivation's identity as an
input to other derivations).

The input-substitution logic is factored out into `maskDerivation`,
which converts a `Derivation` into a `DerivationT<HashModuloInputs>`,
and a second `pathDerivationModulo` overload that handles the per-input
hash-modulo lookup and map insertion.

`HashModuloInputs` keys the input derivations by `Hash` in a `std::map`.
`Hash::operator<=>` compares bytes left-to-right, which matches
base16-lexicographic order (hex encoding is monotonic per byte), so the
map produces the correct ATerm key ordering directly. Conventiently,
`HashModuloInputs` is just an implementation detail confined
`derivations.cc`

Side note: `hasDynamicDrvDep` uses the range-based APIs now, which is
nice. It is also used one more time in a place we just forgot to
deduplicate before.

Ericson2314 committed:
libstore: assert unreachable output types in masked `unparse`

`unparse` is only called with `maskOutputs` set when computing the hash
modulo of a derivation, and `hashDerivationModulo` returns early for
all content-addressing derivations — with the fixed output hashes for
fixed CA derivations, and with `DeferredDrv` for floating CA and impure
ones — before ever serialising them. So the CA fixed, CA floating, and
impure output types can never appear in a masked `unparse`. Add
`panic()` assertions to document and enforce this invariant.

This also removes the now-dead `maskOutputs` ternary from the `CAFixed`
branch, since only the non-masked path can reach it.

Ericson2314 committed:
Add an `Output` type parameter to derivation

Right now it is just noise, but it will become useful in the next
commit.

Co-authored-by: Amaan Qureshi <git@amaanq.com>
Assisted-by: Claude Fable 5

Ericson2314 committed:
libstore: use output type parameter to eliminate `maskOutputs`

The overall goal here is to tighten up the very complicated hash
derivation modulo logic by making things more type-directed, removing
illegal states and boolean blindness.

To that end, we've done a number of refactors, and now things are
significantly clearer:

- `unparseDerivation` has no masking logic — it delegates to overloaded
  `unparseOutput` functions (one per `DerivationOutput` variant), so the
  rendering is purely determined by the output type.

- `derivationModulo` replaces `maskDerivation` and is similarly
  unopinionated, it *just* converts (by value) a derivation with input
  derivations specified by drv path into one with input derivations
  specified by hash. In other words, it just handles the inductive case
  for the "derivation hash modulo" algorithm, and is really just a
  slight special case of "map inputs".

The magic now all happens in the callers of `derivationModulo`, which
instead of a single function with a constexpr bool is split into two:

- `hashDerivationModulo` (private): This is for actually computing input
  addressed cases, and is only an inductive case (final step actually,
  not the base case, which is the first step).

  Returns `std::optional<Hash>` (nullopt if deferred). It works by
  converting outputs to `Deferred` (they must already be input-addressed
  or deferred), blanking output env vars, then calling
  `derivationModulo`, and finally hashing that derivation modulo. Used by
  `processDerivationOutputPaths` to compute a derivation's own output
  paths. The `Deferred` output type ensures the hash does not depend on
  output paths (avoiding circularity).

- `hashInputDerivationModulo` (public): This is the base case (fixed
  output derivation) or inductive case (hash of input-address with
  already-filled-in output paths).

  Returns `DrvHashModulo`. Handles fixed-CA (returns `CaOutputHashes`),
  extracts `InputAddressed` outputs (returns `DeferredDrv` if any output
  isn't `InputAddressed`), then calls `derivationModulo`. Used for
  computing a derivation's identity as an input to other derivations.
  The `InputAddressed` output type preserves output paths in the hash.

In both cases, the output type check and conversion happen in a single
`std::visit` over the output variants before calling `derivationModulo`,
making it impossible to reach it with an output variant that doesn't
match the target type.

`processDerivationOutputPaths` now works with `std::optional<Hash>`
instead of `DrvHashModulo`, since `CaOutputHashes` can never appear in
this context. We had an assert before but now we use the narrow return
type of `hashDerivationModulo` to prove that the case cannot happen.

Co-authored-by: Amaan Qureshi <git@amaanq.com>
Assisted-by: Claude Fable 5

Ericson2314 committed:
Move the derivation json code to a new file

The single `derivation.cc` was getting too big!

Nothing else is changed, this is a pure move.

Ericson2314 committed:
Pull `FullInputs` into its own header

Code using a `std::set<something>` inputs type should not need to bother
with this, and I would like to increase the amount of code which indeed
doesn't bother with `FullInput` over time.

Ericson2314 committed:
Split some code to new files

Move the derivation ATerm code to new files
-------------------------------------------

Like the earlier move of the JSON code to `derivation-json.cc`: the
single `derivations.{hh,cc}` was getting too big.

This moves the ATerm parser (`parseDerivation` and friends), the ATerm
printer (`unparse` and friends), the hash-modulo machinery (which
operates on ATerm-family intermediate representations), and the
wire-protocol (de)serialization of `BasicDerivation` (which also uses
the legacy ATerm-style representation) to a new
`derivation/aterm.{hh,cc}` pair, with the header taking over the
corresponding declarations from `derivations.hh`.

Almost nothing is changed, this is nearly a pure move. The exceptions:

* `hashDerivationModulo` (the output-masking variant, used by
  `processDerivationOutputPaths` which stays in `derivations.cc`) is no
  longer `static`, and now declared in the new header.

* Users of the moved declarations gain an include of the new header.

Split out "derivation/elaboration" from "derivation-options"
------------------------------------------------------------

Once Derivation owns the options in the next commit, the "elaboration"
code here will only be relevant for converting between the full-featured
in-memory derivation and the on-disk ATerm format. This split separates
it out ahead of that change.

Split out `derivation/output.{hh,cc}` from `derivations.{hh,cc}`
----------------------------------------------------------------
`DerivationOutput` (and the related maps/aliases) get their own header
and source file, `derivation/output.{hh,cc}`. This is another pure move
to make `derivations.{hh,cc}` smaller; the next commit will grow the
definitions around the outputs map, so having the output type separate
first keeps that diff focused.

Ericson2314 committed:
Move `OutputChecks` out of `DerivationOptions`

Instead, it is in the new `nix::derivation` namespace.

This commit is the purest of refactors --- absolutely nothing is changed
by identifiers and definition membership. But this churn now will reduce
churn in a future commit.

jaen commented on src/libstore/derivation/aterm.cc:
```diff
@@ -697,37 +719,76 @@ static std::string unparseDerivation(const StoreDirConfig & store, const Derivat
+       environment gets them back. This is the case for all derivations
+       whose options actually stem from the legacy environment-variable
+       encoding. */
+    if (lowered.elaborate(store, drv.name) != drv)
```
Any ideas on how to get rid of this check (or make it faster) while retaining correctness? Reading the inital magic bytes of ATerm? Storing a flag? Something else?

I'm asking because on my Chromimum workload (mock builders, so mostly measures Nix overhead) I wanted to reduce parsing overhead by doing it lazily, then noticed this PR and wanted to adapt to it, maybe try CBOR or Borsh for faster parsing, but it turns out this forced my lazy accessor an caused it to perform even worse ("check guarded" is it being `ifdef`'d out to check the overhead):

```
┌──────────────────────────────┬─────────────┬─────────────┐
│            config            │ time to 400 │  tail rate  │
├──────────────────────────────┼─────────────┼─────────────┤
│ baseline, no PR (lazy)       │      ~560 s │ 6.2 s/chunk │
├──────────────────────────────┼─────────────┼─────────────┤
│ PR + round-trip check + lazy │      ~700 s │ 8.6 s/chunk │
├──────────────────────────────┼─────────────┼─────────────┤
│ PR + check guarded + eager   │       786 s │ 9.5 s/chunk │
├──────────────────────────────┼─────────────┼─────────────┤
│ PR + check guarded + lazy    │       445 s │ 4.0 s/chunk │
└──────────────────────────────┴─────────────┴─────────────┘
```

Now, this workload is rather pathological, because some derivations end up being tens of megabytes in size (the biggest is 35 MB I think) and I'm gonna try to get that down on my end as well, but I figured it'd be a good stress test, so I'm trying to see what else breaks at this scale.

Ericson2314 replied:
Yeah we can mechanically optimize this, and test it against this the "naive implementation". It should be a great thing to fuzz!
